### PR TITLE
HPCC-35787 Avoid rerunning tests on an empty force push

### PR DESCRIPTION
**Pull request by Michael-Gardner**
## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [ ] This change is a new feature (non-breaking change which adds functionality).
- [x] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [x] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [x] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [x] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Cloud-compatibility
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [ ] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [ ] Send notifications about my Pull Request position in Smoketest queue.
- [ ] Test my draft Pull Request.

## Testing:

**Purpose:** Avoid rerunning tests on an empty force push

This commit implements an optimization to prevent GitHub Actions workflows from running unnecessarily when a pull request is synchronized without any actual code changes (e.g., empty force pushes, metadata-only updates).

**Key Changes:**

1. **Added precheck-branch-changes Job** (to 5 workflow files)
   - New job added to: `build-vcpkg.yml`, `codeql-eclwatch.yml`, `test-eclhelper.yml`, `test-helm.yml`, `test-hyperlinks.yml`
   - This job runs before all other jobs to determine if the branch actually has code changes

2. **Three-Step Detection Strategy:**
   - **Fetch Previous Commit** — Retrieves the old commit tip for comparison (handles force-push scenarios)
   - **Check Previous Run Status** — Uses GitHub API to check if the previous workflow run for the old commit failed/was cancelled/timed out
   - **Compare Changes** — Uses git diff to compare old and new commit SHA to detect actual code changes

3. **Smart Logic:**
   - Skips workflows if no code changes detected AND previous run was successful
   - Re-runs workflows if no code changes detected BUT previous run failed (allows retry without code changes)
   - Always runs for new PRs (opened action) or non-PR events
   - Handles edge cases like timed_out workflows

4. **Job Dependencies Updated:**
   - Modified all build/test jobs to depend on `precheck-branch-changes`
   - Added condition: `needs.precheck-branch-changes.outputs.branch_changed == 'true'`
   - This ensures jobs only run when there are actual changes or when needed

5. **Workflow-Specific Configuration:**
   - Each workflow has its filename hardcoded (build-vcpkg.yml, test-helm.yml, etc.) for the GitHub API call

**Impact:**
- Saves CI/CD resources by skipping expensive build/test jobs on metadata-only commits
- Preserves reliability by re-running if previous attempt failed
- Maintains correctness by always running on actual code changes

**Tested on:** [michael-gardner/HPCC-Platform/actions](https://github.com/Michael-Gardner/HPCC-Platform/actions)

### CODE REVIEW

**Review by GordonSmith — CHANGES_REQUESTED** [review]
Can you address the copilot comments and I will review then.

File: .github/workflows/build-vcpkg.yml [hunk 1]
```diff
@@ -60,6 +60,72 @@ permissions:
...
+              if [[ "$PREV_FAILED" == "true" ]]; then
+                echo "No code changes detected, but previous run failed. Re-Running."
+                echo "branch_changed=true" >> $GITHUB_OUTPUT
+              else
+                echo "branch_changed=false" >> $GITHUB_OUTPUT
+                echo "No code changes detected (metadata only). Skipping build and tests."
+              fi
+            else
+              echo "branch_changed=true" >> $GITHUB_OUTPUT
+              echo "Branches have changes."
```
[the remark this answers is not in the document]
Comment (Michael-Gardner) [reply .github/workflows/build-vcpkg.yml]: The github.event.action == synchonize can only ever be set on a github.event_name pull_request. It's guaranteed to have github.event.before and github.event.after.  The only case where a github.even.before wouldn't be present is on an open pull request which would be a different github.event.action == opened.  Going to disregard this suggestion.

File: .github/workflows/test-hyperlinks.yml [hunk 2]
```diff
@@ -56,8 +56,80 @@ concurrency:
   group: ${{ github.workflow }}-${{ github.ref }}
   cancel-in-progress: true
 
+permissions:
+  contents: read
+  actions: write
```
[the remark this answers is not in the document]
Comment (Michael-Gardner) [reply .github/workflows/test-hyperlinks.yml:61]: Upload artifacts at the end of the file needs to write.

**Comment by Michael-Gardner** [comment]
@GordonSmith I fixed all the copilot issues and requested a new review from you.  I noticed after pushing my changes that the precheck-branch-changes was returning "Previous run conclusion: none" which was in error. I think it was a pagination issue with the api request. This repo has a considerable higher amount of traffic than my personal repo that I tested on. I'm pushing a fix and will have everything ready for you to look at in the morning your time.

**Review by GordonSmith — CHANGES_REQUESTED** [review]
Looks good - can you refactor the common code into a new action `.github/actions/branch-changes/action.yaml` (or similar) and reuse?

**Comment by Michael-Gardner** [comment]
> Looks good - can you refactor the common code into a new action `.github/actions/branch-changes/action.yaml` (or similar) and reuse?

Possibly. It depends if the github.event context doesn't change.  I'll check.

**Comment by Michael-Gardner** [comment]
@GordonSmith back to you. Moved the logic into its own action like you suggested.

**Review by GordonSmith — APPROVED** [review]

**Review by ghalliday — APPROVED** [review]
@Michael-Gardner please squash

**Comment by Michael-Gardner** [comment]
@ghalliday Squashed and should be good for merging.
